feat(node): restrict replication fetch range when node is full

maqi opened the pull request:
## Description

When node is full, only fetch new copies that is closer than the current farthest record.

reviewpad:summary

maqi committed:
feat(node): restrict replication fetch range when node is full

JasonPaulGithub reviewed (commented):
the code looks fine but what's going on here may need another pair of eyes

joshuef committed:
feat(networking): clear  record on valid put

If we're able to store data agin, we clear the farthest distance in the replication fetcher.
(The assumed sitation is that the network has grown, and we're not responsible for less data).
